x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepository.java

```diff
@@ -6,11 +6,682 @@
+import org.elasticsearch.repositories.RepositoryData;
+import org.elasticsearch.repositories.RepositoryException;
+import org.elasticsearch.repositories.RepositoryStats;
+import org.elasticsearch.repositories.ShardGenerations;
+import org.elasticsearch.repositories.blobstore.BlobStoreRepository;
+import org.elasticsearch.snapshots.SnapshotId;
+import org.elasticsearch.snapshots.SnapshotInfo;
+import org.elasticsearch.snapshots.SnapshotShardFailure;
+
+import javax.crypto.KeyGenerator;
+import javax.crypto.SecretKey;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.NoSuchFileException;
+import java.security.GeneralSecurityException;
+import java.security.SecureRandom;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+public class EncryptedRepository extends BlobStoreRepository {
+    static final Logger logger = LogManager.getLogger(EncryptedRepository.class);
+    // the following constants are fixed by definition
     static final int GCM_TAG_LENGTH_IN_BYTES = 16;
     static final int GCM_IV_LENGTH_IN_BYTES = 12;
-    static final int AES_BLOCK_SIZE_IN_BYTES = 128;
+    static final int AES_BLOCK_LENGTH_IN_BYTES = 128;
+    // the following constants require careful thought before changing because they will break backwards compatibility
     static final String DATA_ENCRYPTION_SCHEME = "AES/GCM/NoPadding";
     static final long PACKET_START_COUNTER = Long.MIN_VALUE;
-    static final int MAX_PACKET_LENGTH_IN_BYTES = 1 << 30;
+    static final int MAX_PACKET_LENGTH_IN_BYTES = 8 << 20; // 8MB
+    // this should be smaller than {@code #MAX_PACKET_LENGTH_IN_BYTES} and it's what {@code EncryptionPacketsInputStream} uses
+    // during encryption and what {@code DecryptionPacketsInputStream} expects during decryption (it is not configurable)
+    static final int PACKET_LENGTH_IN_BYTES = 64 * (1 << 10); // 64KB
+    // the path of the blob container holding all the DEKs
+    // this is relative to the root base path holding the encrypted blobs (i.e. the repository root base path)
+    static final String DEK_ROOT_CONTAINER = ".encryption-metadata"; // package private for tests
+    static final int DEK_ID_LENGTH = 22; // {@code org.elasticsearch.common.UUIDS} length
+
+    // the following constants can be changed freely
+    private static final String RAND_ALGO = "SHA1PRNG";
+    // each snapshot metadata contains an unforgeable identifier of the repository password of the master node that started the snapshot
+    // this hash is then verified on each data node before the actual shard files snapshot, as well as on the
+    // master node that finalizes the snapshot (could be a different master node, if a master failover occurred during the snapshot)
```

Comment:
You understood correctly, and indeed I've explained more of the how instead of the why.

I've changed the javadoc here to say more about the why, similar to `#validateLocalRepositorySecret` (I don't think the repetition is really detrimental).